{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# apytram v1.0\n",
    "## Automated ( PYthon implemented ) Target Restricted Assembly Method"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Contact\n",
    "\n",
    "carine.rey@ens-lyon.fr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preamble\n",
    "\n",
    "This software is inspired from aTRAM (see References) but was implemented in Python and internal strategies have been designed differently. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Why use apytram?\n",
    "\n",
    "apytram allows assembling sequences from RNA-seq data (paired-end or not) using a reference homologous sequence, possibly coming from a different species."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How does it work?\n",
    "\n",
    "\n",
    "### Database building\n",
    "The RNA-seq data (a fastq or fasta file), given by the -fq or -fa option, is formatted into a BLAST database, whose  name is given by the -d option.\n",
    "\n",
    "The data type, paired or single end, must be given by the -dt option. If data are paired-end, all reads (1 and 2) must be concatenated in a uniq file. WARNING: Paired read names must end with 1 or 2.\n",
    "Reads contained in this file will all be used, so the file must have already been clean up.\n",
    "\n",
    "If this file is a fastq file, it will be converted into a fasta file. This conversion can take a lot of time (between XX and YY hours on a laptop). \n",
    "\n",
    "As the database building step is time consuming, if a BLAST database is already present, the database building will be skipped.\n",
    "\n",
    "### Iterative process\n",
    "\n",
    "The sequence in the query file (-q option) will serve as the first bait sequence. At this time, only one reference sequence must be present in the query file. The possibility to use a multi-reference query file is in development.\n",
    "\n",
    "#### A classical iteration (see Figure XX)\n",
    "*    Reads recruting (Blast):\n",
    "\n",
    "     Bait sequences are used to recrute homologous reads by BLAST. The -e option allows fixing the evalue threshold.\n",
    "     If the data is paired-end (-dt option), all paired reads are added to the read list.\n",
    "     \n",
    "     Sequences of all these reads are put in a temporary file which is accesible if the -tmp option is used.\n",
    "     \n",
    "     \n",
    "*    Read assembly (Trinity):\n",
    "\n",
    "     All reads present in this temporary file are assembled (de novo) by Trinity with default parameters. If the data contains paired-end reads, Trinity takes reads as paired reads.\n",
    "     \n",
    "     \n",
    "*    Reconstructed sequences quality (Exonerate):\n",
    "\n",
    "     Exonerate is used to align each sequence to the reference. The length of the sequence, the length of the alignment, the percentage of identity, the alignment score are collected.\n",
    "     \n",
    "     \n",
    "*    Reconstructed sequence filtering:\n",
    "\n",
    "     Reconstructed sequences are filtered according to the option -id -mal -len. Sequences which pass all filters are used as bait sequences for the next iteration.\n",
    "     \n",
    "     \n",
    "*    Comparaison with previous iteration  (Exonerate):\n",
    "     \n",
    "     Exonerate is used to find the \"parent\" contig from the previous contig for each contig and check they are really different.\n",
    "     \n",
    "\n",
    "*    Coverage calculation (Mafft):\n",
    "\n",
    "    apytram calculates 2 coverage values, a Strictcoverage that means the percentage of sites of the query with a homologous site in the reconstructed sequences, and a Largecoverage that means the percentage of sites in the alignment with at least one representant in the reconstructed sequences divided by the length of the reference so by definition it can be superior to 100.\n",
    "    \n",
    "![ Figure to explain.](https://github.com/CarineRey/apytram/blob/master/Documentation/CoverageExplanation.png?raw=true \"Coverage explain\")\n",
    "\n",
    "At the end of an iteration, reconstructed sequences become the bait sequences for the next iteration.\n",
    "\n",
    "\n",
    "#### Criteria to stop the iterative process\n",
    "If one of these criteria is completed during an iteration, the iterative process will stop.\n",
    "The following criteria are implemented as default settings : \n",
    "*    The number of max iteration (-i option) is attained.\n",
    "*    Recruted reads are the same as during the previous iteration.\n",
    "*    Reconstructed sequences at the end of an iteration are almost the same as after the previous iteration. That means each sequence of an iteration has a corresponding sequence in the previous iteration with at least 99% identity and 98% of its length.\n",
    "*    The number of reconstructed sequences has not changed AND the total length, score and LargeCoverage of all reconstructed sequences have not been improved.\n",
    "\n",
    "If the --required_coverage option is used, the iterating process will stop if the Strict coverage is superior to the Required_coverage.\n",
    "\n",
    "All these criteria are not applied if the --finish_all_iteration is used.\n",
    "\n",
    "### Final Filter\n",
    "\n",
    "A final filter (-fmal -fid -flen option) can be applied on the reconstructed sequences to be more stringent than the threshold used during the iterative process. \n",
    "\n",
    "### Writing outputs file\n",
    "Finally, output files are writen:\n",
    "*    `$OUTPUT_PREFIX`**.fasta**: \n",
    "\n",
    "     A fasta file containing all reconstructed sequences of the last iteration which pass the final filter.\n",
    "\n",
    "\n",
    "*    `$OUTPUT_PREFIX`**.best.fasta**:\n",
    "\n",
    "     A fasta file containing the reconstructed sequences of the last iteration with the best scores, and which pass the final filter.\n",
    "\n",
    "\n",
    "*    `$OUTPUT_PREFIX`**.stats.pdf**:\n",
    "    \n",
    "    available if --plot option. A pdf containing 2 figures to know global information at each iteration.\n",
    "\n",
    "\n",
    "*    `$OUTPUT_PREFIX`**.stats.csv**:\n",
    "      \n",
    "      available if --plot or --stats options. A csv file containing the raw data needed to draw the figures present in the .stats.pdf \n",
    "\n",
    "\n",
    "*    `$OUTPUT_PREFIX`**.ali.png**:\n",
    "\n",
    "        A figure representing the alignment of all reconstructed sequences which pass the final filter on the query. White represents a gap, blue a base of the reference, green an identical base of the reference in a reconstructed sequence, red a different base of the reference and yellow a base corresponding to a gap in the reference."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dependencies\n",
    "### Software (Must be available in the \\$PATH)\n",
    "\n",
    "*    Trinity >=v2.1 [Samtools  = v0.19, Java 1.7, bowtie] ([Download Trinity here](https://github.com/trinityrnaseq/trinityrnaseq/releases)) ([Download Samtools here](https://sourceforge.net/projects/samtools/files/samtools/0.1.19/)) \n",
    "    \n",
    "*    Mafft >=v7 ([Download Mafft here](https://http://mafft.cbrc.jp/alignment/software/))\n",
    "\n",
    "*    exonerate ([Download exonerate here](https://github.com/nathanweeks/exonerate))\n",
    "\n",
    "*    Blastplus\n",
    "\n",
    "To add a software to the \\$PATH environment variable:\n",
    "```sh\n",
    "PATH=/path/to/the/software:$PATH\n",
    "```\n",
    "\n",
    "You can also install this software via a package management program (e.g. aptitude, synaptic ...).\n",
    "\n",
    "For example to install exonerate:\n",
    "```bash\n",
    "apt-get install exonerate\n",
    "```\n",
    "### Python library\n",
    "*    pandas\n",
    "*    Matplolib >= 1.13\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How to get apytram?\n",
    "\n",
    "### By downloading\n",
    "apytram is available on github [https://github.com/CarineRey/apytram](https://github.com/CarineRey/apytram).\n",
    "\n",
    "### By command line\n",
    "This command will create a directory named \"apytram\".\n",
    "```sh\n",
    "git clone https://github.com/CarineRey/apytram.git\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basic use\n",
    "\n",
    "### Using the included example (Nucleotide query)\n",
    "This example is included in the apytram directory. It should run in about a dozen seconds.\n",
    "Code must be executed in the parent directory of apytram. Before running this example, check that all dependencies are checked.\n",
    "\n",
    "```bash\n",
    "export OUT=\"exec_example\"\n",
    "apytram/apytram.py -d $OUT/db/examplefq \\\n",
    "             -dt paired \\\n",
    "             -fq apytram/example/example_db.fastq \\\n",
    "             -q apytram/example/ref_gene.fasta \\\n",
    "             -out $OUT/apytram \\\n",
    "             -log $OUT/apytram.log \\\n",
    "             --plot \\\n",
    "             --plot_ali \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How to get help?\n",
    "```sh\n",
    "/apytram_path/apytram.py -h\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simplest use\n",
    "```sh\n",
    "./apytram.py -d $db_name -t $db_type -fq $fastq_name -q $query_file_name\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Change the number of threads\n",
    "```sh\n",
    "./apytram.py -d $db_name -t $db_type -fq $fastq_name -q $query_file_name -threads $nb_threads\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Access to tempory files\n",
    "```sh\n",
    "./apytram.py -d $db_name -t $db_type -fq $fastq_name -q $query_file_name -tmp $tmp_directory_name\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Database building only\n",
    "```sh\n",
    "./apytram.py -d $db_name -t $db_type -fq $fastq_name\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use of the final filter\n",
    "*    If you want to keep only reconstructed sequences which have a length superior to **X percent** of the reference sequence:\n",
    "    ```sh\n",
    "    ./apytram.py -d $db_name -t $db_type -fq $fastq_name -q $query_file_name -flen X\n",
    "    ```\n",
    "    \n",
    "*    If you want to keep only reconstructed sequences which have a identity percentage superior to Y with the reference sequence on the length of their alignment:\n",
    "    ```sh\n",
    "    ./apytram.py -d $db_name -t $db_type -fq $fastq_name -q $query_file_name -fid Y\n",
    "    ```\n",
    "    \n",
    "*    If you want to keep only reconstructed sequences which align on the reference on a length superior to **Z percent** of the reference sequence:\n",
    "    ```sh\n",
    "    ./apytram.py -d $db_name -t $db_type -fq $fastq_name -q $query_file_name -fmal Z\n",
    "    \n",
    "    ```\n",
    "    \n",
    "*   If you want to combine all these things:\n",
    "    ```sh\n",
    "    ./apytram.py -d $db_name -t $db_type -fq $fastq_name -q $query_file_name -flen X -fid Y -fmal Z\n",
    "    \n",
    "    ```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Speed optimization\n",
    "\n",
    "Paired-end RNA-seq data run faster than single-end data.\n",
    "\n",
    "To save time:\n",
    "*     not use -tmp option\n",
    "*     not use --keep_iterations\n",
    "*     not use --finish_all_iter\n",
    "*     not use the --plot and --plot_ali options\n",
    "\n",
    "If you want to use apytram on several query files and you have several available threads, it is more efficient to minimize the number of threads by apytram job than to maximize the number of threads by job. Each job will be slower but at the end you will save time. This is due to the non linearity of the time saved by Trinity and Blast when the number of threads is increased."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Optimizing the accuracy\n",
    "\n",
    "See folowing options:\n",
    "*    -e\n",
    "*    -mal\n",
    "*    -id\n",
    "*    -len\n",
    "*    -fid\n",
    "*    -fmal\n",
    "*    -flen"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Following the program's progress\n",
    "\n",
    "A job can take some minutes to several hours to complete. To know the progress of your job you can look into the log file (-log option or by default apytram.log).\n",
    "\n",
    "You can look at the `$OUTPUT_PREFIX.stats.pdf` (--plot option) at the end the job, to have general information on the progress of your job. All values needed to create the plot in `$OUTPUT_PREFFIX.stats.pdf` are available in `$OUTPUT_PREFFIX.stats.csv`. The `$OUTPUT_PREFFIX.stats.csv` file can be only created using the --stats option."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Advanced use\n",
    "For the moment, it is the help message.\n",
    "\n",
    "```\n",
    "usage: apytram.py [-h] [--version] -d DATABASE -dt {single,paired} [-fa FASTA]\n",
    "                  [-fq FASTQ] [-q QUERY] [-pep QUERY_PEP] [-i ITERATION_MAX]\n",
    "                  [-out OUTPUT_PREFIX] [-log LOG] [-tmp TMP]\n",
    "                  [--keep_iterations] [--no_best_file] [--no_last_iter_file]\n",
    "                  [--stats] [--plot] [--plot_ali] [-e EVALUE] [-id MIN_ID]\n",
    "                  [-mal MIN_ALI_LEN] [-len MIN_LEN]\n",
    "                  [--required_coverage REQUIRED_COVERAGE] [--finish_all_iter]\n",
    "                  [-flen FINAL_MIN_LEN] [-fid FINAL_MIN_ID]\n",
    "                  [-fmal FINAL_MIN_ALI_LEN] [-threads THREADS]\n",
    "\n",
    "Run apytram.py on a fastq file to retrieve homologous sequences of bait\n",
    "sequences.\n",
    "\n",
    "optional arguments:\n",
    "  -h, --help            show this help message and exit\n",
    "  --version             show program's version number and exit\n",
    "\n",
    "Required arguments:\n",
    "  -d DATABASE, --database DATABASE\n",
    "                        Database prefix name. If a database with the same name\n",
    "                        already exists, the existing database will be kept and\n",
    "                        the database will NOT be rebuilt.\n",
    "  -dt {single,paired}, --database_type {single,paired}\n",
    "                        single or paired end RNA-seq data. WARNING: Paired\n",
    "                        read names must finished by 1 or 2.\n",
    "\n",
    "Input Files:\n",
    "  -fa FASTA, --fasta FASTA\n",
    "                        Fasta formated RNA-seq data to build the database of\n",
    "                        reads\n",
    "  -fq FASTQ, --fastq FASTQ\n",
    "                        Fastq formated RNA-seq data to build the database of\n",
    "                        reads. (The fastq will be first converted to a fasta\n",
    "                        file. This process can require some time.\n",
    "  -q QUERY, --query QUERY\n",
    "                        Fasta file (nucl) with bait sequences for the apytram\n",
    "                        run. If no query is submitted, the program will just\n",
    "                        build the database.\n",
    "  -pep QUERY_PEP, --query_pep QUERY_PEP\n",
    "                        Fasta file containing the query in the peptide format.\n",
    "                        It will be used at the first iteration as bait\n",
    "                        sequences to fish reads. It is compulsory to include\n",
    "                        also the query in nucleotide format (-q option)\n",
    "  -i ITERATION_MAX, --iteration_max ITERATION_MAX\n",
    "                        Maximum number of iterations. (Default 5)\n",
    "\n",
    "Output Files:\n",
    "  -out OUTPUT_PREFIX, --output_prefix OUTPUT_PREFIX\n",
    "                        Output prefix (Default ./apytram)\n",
    "  -log LOG              a log file to report avancement (default: apytram.log)\n",
    "  -tmp TMP              Directory to stock all intermediary files for the\n",
    "                        apytram run. (default: a directory in /tmp which will\n",
    "                        be removed at the end)\n",
    "  --keep_iterations     A fasta file will be created at each iteration.\n",
    "                        (default: False)\n",
    "  --no_best_file        By default, a fasta file (Outprefix.best.fasta)\n",
    "                        containing only the best sequence is created. If this\n",
    "                        option is used, it will NOT be created.\n",
    "  --no_last_iter_file   By default, a fasta file (Outprefix.fasta) containing\n",
    "                        all sequences from the last iteration is created. If\n",
    "                        this option is used, it will NOT be created.\n",
    "  --stats               Create files with statistics on each iteration.\n",
    "                        (default: False)\n",
    "  --plot                Create plots to represent the statistics on each\n",
    "                        iteration. (default: False)\n",
    "  --plot_ali            Create file with a plot representing the alignement of\n",
    "                        all sequences from the last iteration on the query\n",
    "                        sequence. Take some seconds. (default: False)\n",
    "\n",
    "Thresholds for EACH ITERATION:\n",
    "  -e EVALUE, --evalue EVALUE\n",
    "                        Evalue threshold of the blastn of the bait queries on\n",
    "                        the database of reads. (Default 1e-3)\n",
    "  -id MIN_ID, --min_id MIN_ID\n",
    "                        Minimum identity percentage of a sequence with a query\n",
    "                        on the length of their alignment so that the sequence\n",
    "                        is kept at the end of a iteration (Default 20)\n",
    "  -mal MIN_ALI_LEN, --min_ali_len MIN_ALI_LEN\n",
    "                        Minimum alignment length of a sequence on a query to\n",
    "                        be kept at the end of a iteration (Default 180)\n",
    "  -len MIN_LEN, --min_len MIN_LEN\n",
    "                        Minimum length to keep a sequence at the end of a\n",
    "                        iteration. (Default 200)\n",
    "\n",
    "Criteria to stop iteration:\n",
    "  --required_coverage REQUIRED_COVERAGE\n",
    "                        Required coverage of a bait sequence to stop iteration\n",
    "                        (Default: No threshold)\n",
    "  --finish_all_iter     By default, iterations are stop if there is no\n",
    "                        improvment, if this option is used apytram will finish\n",
    "                        all iteration (-i).\n",
    "\n",
    "Thresholds for Final output files:\n",
    "  -flen FINAL_MIN_LEN, --final_min_len FINAL_MIN_LEN\n",
    "                        Minimum PERCENTAGE of the query length to keep a\n",
    "                        sequence at the end of the run. (Default: 0)\n",
    "  -fid FINAL_MIN_ID, --final_min_id FINAL_MIN_ID\n",
    "                        Minimum identity PERCENTAGE of a sequence with a query\n",
    "                        on the length of their alignment so that the sequence\n",
    "                        is kept at the end of the run (Default 0)\n",
    "  -fmal FINAL_MIN_ALI_LEN, --final_min_ali_len FINAL_MIN_ALI_LEN\n",
    "                        Alignment length between a sequence and a query must\n",
    "                        be at least this PERCENTAGE of the query length to\n",
    "                        keep this sequence at the end of the run. (Default: 0)\n",
    "\n",
    "Miscellaneous options:\n",
    "  -threads THREADS      Number of available threads. (Default 1)\n",
    "  \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reference\n",
    "Allen, JM, DI Huang, QC Cronk, KP Johnson. 2015. aTRAM automated target restricted assembly method a fast method for assembling loci across divergent taxa from next-generation sequencing data. BMC Bioinformatics 16:98 DOI 10.1186/s12859-015-0515-2"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
